### PR TITLE
Add Mixture of Tokens model

### LINKED ISSUE #28285: Add Mixture of Tokens model
### Model description

Mixture of Tokens is a new architecture / technique proposed in [Mixture of Tokens: Efficient LLMs through Cross-Example Aggregation](https://arxiv.org/abs/2310.15961) and accompanying [blog](https://llm-random.github.io/posts/mixture_of_tokens/) by Szymon Antoniak, Sebastian Jaszczur et al.

It builds on expert-choice MoE, aggregating across sequences in a batch rather than positions in a sequence, and doing so in a continuous fashion. It promises efficiency and greater training stability than vanilla MoE.

In collaboration with the authors, we (me + 3 others) would like to add an implementation matching the architecture from the paper to HF transformers and later publish corresponding checkpoints. We believe this will make it significantly easier for the community to experiment with this approach, as the original implementation is quite dense and contained in an active research repo.

We believe a good approach is to start from the GPT2 HF model. We will have the assistance of the original authors for making sure the details match.

Please advise if you have any suggestions at this stage, and also what kinds of tests you would like to see in the finalized implementation. 
As we understand, the next step is for us to create a template with https://github.com/huggingface/transformers/tree/main/templates/adding_a_new_model and get coding.

### Open source status

- [X] The model implementation is available
- [ ] The model weights are available

### Provide useful links for the implementation

https://github.com/llm-random/llm-random

https://github.com/sebastianjaszczur

### PR DESCRIPTION
**Pull request by szysad**
# What does this PR do?

Fixes https://github.com/huggingface/transformers/issues/28285

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Review by ArthurZucker — COMMENTED** [review]
Hey! Would recommend you to start by [sharing the model ](https://huggingface.co/docs/transformers/en/model_sharing)on the hub! Also filling an issue with the checkpoints, the paper, the code etc! That way we can track the interest from the community. 

Make sure to link it to this PR!

**Review by szysad — COMMENTED** [review]
what are experts in mot setup?

**Comment by JakimPL** [comment]
@ArthurZucker We consider our work done, and we are ready for the review :). Failed tests seem unrelated to this code.

**Comment by chedatomasz** [comment]
I performed some additional cleanup and extended the tests to cover additional cases.
The issue with the paper and the original code is available at https://github.com/huggingface/transformers/issues/28285).
The model weights are available at https://huggingface.co/jaszczur/mixture_of_tokens. This is a previously unreleased checkpoint created in cooperation with the paper authors.
We updated the documentation with our changes.
The test failures are flaky tests unrelated to our changes.
Please review and advise whether this looks good.

File: src/transformers/models/mot/modeling_mot.py [hunk 1]
```diff
@@ -0,0 +1,1621 @@
+# coding=utf-8
+# Copyright 2024 The OpenAI Team Authors and HuggingFace Inc. team.
```
Comment (ArthurZucker) [inline src/transformers/models/mot/modeling_mot.py:2]:
```suggestion
# Copyright 2024  HuggingFace Inc. team.
```

File: src/transformers/models/mot/modeling_mot.py [hunk 2]
```diff
@@ -0,0 +1,1621 @@
...
+            else:
+                x = forward_fn(self, x)
+                return x[:size, :, :]
+        else:
+            return forward_fn(self, x)
+
+    return _forward
+
+
+# Copied from transformers.models.gpt2.modeling_gpt2.GPT2Attention with GPT2->MoT
```
Comment (ArthurZucker) [inline src/transformers/models/mot/modeling_mot.py:78]: Arf, GPT2 is a super old model, most probably not a good reference implementation. Lots of issues with the conv1d have been reported 😢

File: src/transformers/models/mot/modeling_mot.py [hunk 3]
```diff
@@ -0,0 +1,1621 @@
...
+        )
+
+        self.controller = nn.Parameter(
+            self.get_init_weight(
+                (self.d_model, self.n_expert),
+                fan_in=self.d_model,
+                init_type=self.init_type,
+                scale=self.init_scale,
+            )
+        )
```
Comment (ArthurZucker) [inline src/transformers/models/mot/modeling_mot.py:362]: we never call init weights in the init of modules, but in the `PreTrainedModel` class !

File: src/transformers/models/mot/modeling_mot.py [hunk 4]
```diff
@@ -0,0 +1,1621 @@
...
+        std = (scale / fan_in) ** 0.5
+        low = -2 * scale
+        high = 2 * scale
+        t = torch.zeros(shape, dtype=dtype)
+        return trunc_normal_(t, mean=0.0, std=std, a=low, b=high)
+
+    @staticmethod
+    def stable_softmax_temperature(x: torch.Tensor, temperature: float, dim: int = -1) -> torch.Tensor:
+        return F.softmax(x / temperature, dim=dim)
+
```
Comment (ArthurZucker) [inline src/transformers/models/mot/modeling_mot.py:398]: same comment about inits

File: src/transformers/models/mot/modeling_mot.py [hunk 5]
```diff
@@ -0,0 +1,1621 @@
...
+        return torch.cat([x, padding_sequences], dim=0)
+
+    @with_batch_size_alignment
+    def forward(self, x):
+        x = self.group_tokens(x)
+        merge_weights, emit_weights = self.calculate_mixed_tokens_with_weights(x)
+        x = self.merge_map_emit(x, merge_weights, emit_weights)
+        x = self.redistribute_tokens(x)
+        x = self.dropout(x)
+        return x
```
Comment (ArthurZucker) [inline src/transformers/models/mot/modeling_mot.py:419]: no single letter variables please!

File: src/transformers/models/mot/modeling_mot.py [hunk 6]
```diff
@@ -0,0 +1,1621 @@
...
+        elif self.sparsity_dim != 1:
+            raise NotImplementedError
+
+        free_dimension = x.size(1)
+        assert (
+            free_dimension % self.group_size == 0
+        ), f"free dimension = {free_dimension} should be divisible by group size = {self.group_size}"
+
+        x = x.view(x.size(0), -1, self.group_size, self.d_model)
+        return x
```
Comment (ArthurZucker) [inline src/transformers/models/mot/modeling_mot.py:450]: let's avoid asserts in the codebase

File: src/transformers/models/mot/modeling_mot.py [hunk 7]
```diff
@@ -0,0 +1,1621 @@
...
+            emit_weights = self.stable_softmax_temperature(merge_logits, temp_emit, dim=emit_softmax_dim)
+        else:
+            emit_weights = merge_weights
+
+        if self.use_discrete_routing:
+            merge_weights = self.argmax_one_hot(merge_weights, dim=merge_softmax_dim)
+            emit_weights = self.argmax_one_hot(emit_weights, dim=emit_softmax_dim)
+        return merge_weights, emit_weights
+
+    def merge_map_emit(self, x, merge_weights, emit_weights):
```
Comment (ArthurZucker) [inline src/transformers/models/mot/modeling_mot.py:497]: same about single letter variables

File: src/transformers/models/mot/modeling_mot.py [hunk 8]
```diff
@@ -0,0 +1,1621 @@
...
+        inner_dim = config.n_inner if config.n_inner is not None else 4 * hidden_size
+        attention_class = MoT_ATTENTION_CLASSES[config._attn_implementation]
+
+        self.ln_1 = nn.LayerNorm(hidden_size, eps=config.layer_norm_epsilon)
+        self.attn = attention_class(config=config, layer_idx=layer_idx)
+        self.ln_2 = nn.LayerNorm(hidden_size, eps=config.layer_norm_epsilon)
+
+        if config.add_cross_attention:
+            self.crossattention = attention_class(config=config, is_cross_attention=True, layer_idx=layer_idx)
+            self.ln_cross_attn = nn.LayerNorm(hidden_size, eps=config.layer_norm_epsilon)
```
Comment (ArthurZucker) [inline src/transformers/models/mot/modeling_mot.py:545]: we usually separate cross attention and self attention only layers

File: src/transformers/models/mot/modeling_mot.py [hunk 9]
```diff
@@ -0,0 +1,1621 @@
...
+        0: [0, 1, 2, 3, 4, 5, 6, 7],
+        1: [8, 9, 10, 11, 12, 13, 14, 15],
+        2: [16, 17, 18, 19, 20, 21, 22, 23],
+        3: [24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35],
+    }
+    model.parallelize(device_map)  # Splits the model across several devices
+    model.deparallelize()  # Put the model back on cpu and cleans memory by calling torch.cuda.empty_cache()
+    ```
+"""
+
```
Comment (ArthurZucker) [inline src/transformers/models/mot/modeling_mot.py:790]: this is deprecated in favor of accelerate!

File: src/transformers/models/mot/modeling_mot.py [hunk 10]
```diff
@@ -0,0 +1,1621 @@
...
+        self.model_parallel = False
+        self.device_map = None
+        self.first_device = "cpu"
+        self.last_device = "cpu"
+        self.wte = self.wte.to("cpu")
+        self.wpe = self.wpe.to("cpu")
+        for index in range(len(self.h)):
+            self.h[index] = self.h[index].to("cpu")
+        self.ln_f = self.ln_f.to("cpu")
+        torch.cuda.empty_cache()
```
Comment (ArthurZucker) [inline src/transformers/models/mot/modeling_mot.py:859]: to remove

File: src/transformers/models/mot/modeling_mot.py [hunk 11]
```diff
@@ -0,0 +1,1621 @@
...
+    @add_start_docstrings(DEPARALLELIZE_DOCSTRING)
+    def deparallelize(self):
+        warnings.warn(
+            "Like `parallelize`, `deparallelize` is deprecated and will be removed in v5 of Transformers.",
+            FutureWarning,
+        )
+        self.transformer.deparallelize()
+        self.transformer = self.transformer.to("cpu")
+        self.lm_head = self.lm_head.to("cpu")
+        self.model_parallel = False
```
Comment (ArthurZucker) [inline src/transformers/models/mot/modeling_mot.py:1118]: same comment

File: src/transformers/models/mot/modeling_mot.py [hunk 12]
```diff
@@ -0,0 +1,1621 @@
...
+    ) -> Tuple[Tuple[torch.Tensor]]:
+        """
+        This function is used to re-order the `past_key_values` cache if [`~PreTrainedModel.beam_search`] or
+        [`~PreTrainedModel.beam_sample`] is called. This is required to match `past_key_values` with the correct
+        beam_idx at every generation step.
+        """
+        return tuple(
+            tuple(past_state.index_select(0, beam_idx.to(past_state.device)) for past_state in layer_past)
+            for layer_past in past_key_values
+        )
```
Comment (ArthurZucker) [inline src/transformers/models/mot/modeling_mot.py:1265]: no longer used as well

File: src/transformers/models/mot/modeling_mot.py [hunk 13]
```diff
@@ -0,0 +1,1621 @@
...
+        logits = self.score(hidden_states)
+
+        if input_ids is not None:
+            batch_size, sequence_length = input_ids.shape[:2]
+        else:
+            batch_size, sequence_length = inputs_embeds.shape[:2]
+
+        assert (
+            self.config.pad_token_id is not None or batch_size == 1
+        ), "Cannot handle batch sizes > 1 if no padding token is defined."
```
Comment (ArthurZucker) [inline src/transformers/models/mot/modeling_mot.py:1358]: let's remove asserts

File: src/transformers/models/mot/modeling_mot.py [hunk 14]
```diff
@@ -0,0 +1,1621 @@
...
+            "Lead",
+            "Lead",
+            "Lead",
+            "Lead",
+            "Lead",
+            "Lead",
+            "Lead",
+            "Lead",
+        ],
+    )
```
Comment (ArthurZucker) [inline src/transformers/models/mot/modeling_mot.py:1461]: we tend to avoid these now!

**Review by ArthurZucker — COMMENTED** [review]
Hey ! Sorry for the late review 🤗 
If you still want to push this, will review again, my main advice is to not use GPT2 as a base implementation, and rather look at Llama (it's fairly similar in general 🤗 )
